### ISSUE TITLE
Persistent session server-side but no cookie in browser (using reverse proxy)

### ISSUE THREAD

**Issue by madziszyn** [expressjs/session#622]
Hey,

Server-side code in development environment: 

```js
app.use(bodyParser.json())

app.use(require('cors')({
  origin: function (origin, callback) {
    callback(null, origin)
  },
  credentials: true
}))

app.use(compression())

app.use(helmet())

var expiryDate = new Date(Date.now() + 60 * 60 * 1000 * 24)
app.use(session({
  name: config.sessionCookieName,
  secret: config.sessionCookieSecret,
  cookie: {
    httpOnly: true,
    expires: expiryDate
  },
  resave: false,
  saveUninitialized: false
}))
```

When in development environment everything works perfectly fine. Session is persistent server-side (by persistent i mean session.id is the same for all requests sent from one browser), response header include set-cookie and browser stores sessionId cookie.
By development environment I mean 
server running on localhost:8081
clieent running on localhost:8080
(on my personal computer)

When testing server in production environment i came across something weird.
Production env:
-windows server R2 on remote machine (accessible via let's say mysub.mydomain.com)
-server running on localhost:8081 (on mentioned before remote machine)
-there is reverse proxy (configured in IIS with URL rewrite) that rewrites mysub.mydomain.com to remote machine's localhost:8081 ip address.
-client still on my pc, running on localhost:8080 (but i changed server URL from http://localhost:8081 to http://mysub.mydomain.com) 

I modified previous code: 
```js
app.set('trust proxy', ['loopback', 'linklocal', 'uniquelocal']) // ADDED
app.use(bodyParser.json())

app.use(require('cors')({
  origin: function (origin, callback) {
    callback(null, origin)
  },
  credentials: true
}))

app.use(compression())

app.use(helmet())

var expiryDate = new Date(Date.now() + 60 * 60 * 1000 * 24)
app.use(session({
  name: config.sessionCookieName,
  secret: config.sessionCookieSecret,
  proxy: true, // ADDED
  cookie: {
    domain: '.mydomain.com', // ADDED
    httpOnly: true,
    expires: expiryDate
  },
  resave: false,
  saveUninitialized: false
}))
```
 
In this scenario ALMOST everything works fine. Session is persistent server-side but set-cookie is absent and therefore browser doesn't store cookie. I have no idea how it works (or seems to work). Can you give me some advice? Maybe my whole approach is wrong.

Thanks in advance.

EDIT.
Session data like session.userRole is also persistent.

**Comment by dougwilson** [comment]
Hi @madziszyn I'm sorry you're having trouble. I've never done a setup like you described, so don't know what the solution would be off-hand. Since the app stand alone is working, it doesn't seem like I would just be able to easily reproduce locally. I would probably say can you use a debugger or a tool like Wireshark to see if the `Set-Cookie` header is coming from the Node.js server, to help determine if it is either the Node.js server just not even sending the header or if the header is getting lost somewhere between the Node.js server and your web browser?

**Comment by madziszyn** [comment]
Thanks for the reply. 
I deployed clinet and server on the same remote machine accessed via "production" domain (previously mentioned mysub.mydomain.com) and installed SSL certificate on the IIS and switched to https communication (client-side API calls to reverse proxy).
I changed my server-side code a bit:

```
const app = express()

app.set('trust proxy', true)

app.use(bodyParser.json())

app.use(cors({
  origin: function (origin, callback) {
    callback(null, origin)
  },
  credentials: true
}))

app.use(compression())

app.use(helmet())

var expiryDate = new Date(Date.now() + 60 * 60 * 1000 * 24)
app.use(session({
  name: config.sessionCookieName,
  secret: config.sessionCookieSecret,
  proxy: true,
  cookie: {
    httpOnly: false,
    expires: expiryDate
  },
  resave: false,
  saveUninitialized: true
}))
```

It works perfectly. Set-cookie header is present and cookies are present in browser.

And yet, the original problem described in my first comment is still actual. I used Wireshark as You said but couldn't deduce anything. It's bugging me but I don't have time to investigate on this now... Maybe in couple of weeks.

Bottom line using https solved the problem.

Side-note (communication between reverse-proxy and node server is via http not https)

**Comment by knoxcard** [comment]
close ticket?

**Comment by dougwilson** [comment]
Yes, I'll close this for now. It can always be reopened if there are details that come to light that this module can act on.
